src/main/java/soot/jimple/toolkits/callgraph/VirtualCalls.java

```diff
@@ -108,6 +108,7 @@ public SootMethod resolveNonSpecial(RefType t, NumberedString subSig, boolean ap
 
     SootMethod m = cls.getMethodUnsafe(subSig);
     if (m != null) {
+    	// Why are we not adding to targets for abstract methods
       if (!m.isAbstract()) {
         ret = m;
```

Comment:
@mbenz89 Can there be an edge from one method to another method within the same abstract class?. I am not sure why are we not adding to targets if a method is abstract.

Reply:
Because an edges always has to point to a method that contains actual code! An non-default interface method or an abstract method, however, cannot contain any code that is executed. Wenn dispatching, we want to find the concrete method that is called at a given callside. This can never be an abstract method?

Important: The condition checks if the _method!_ is abstract not the class. We can of course have edges to methods in abstract classes as long as they contain an implementation. Not if the method is abstract still.

Reply:
@mbenz89 Thanks for the clarification. The default implementation is allowing edges between two abstract methods. So I guess a check needs to be added in FastHierarchy to check if the method is abstract before adding to targets.

Reply:
Oh, that should not be possible. Can you show me the specific code? I thought that would only be possible if `allowAbstract` is set to true. This should only be the case if `FastHierarchy.resolveMethod` is called from `SootMethodRefImpl`. Isn't that correct? Is this consistent between default and develop?

Reply:
@mbenz89 There is a check to abstract method in FastHierarchy at https://github.com/Sable/soot/blob/feature/default-methods/src/main/java/soot/FastHierarchy.java#L859. But still we are returning the candidate at FastHierarchy which is the reason getting added to targets at https://github.com/Sable/soot/blob/feature/default-methods/src/main/java/soot/FastHierarchy.java#L884.

Reply:
How is this different to https://github.com/Sable/soot/blob/fcc380a506258291d6fe6954425586a4ae4968ea/src/main/java/soot/jimple/toolkits/callgraph/VirtualCalls.java#L112 and https://github.com/Sable/soot/blob/fcc380a506258291d6fe6954425586a4ae4968ea/src/main/java/soot/jimple/toolkits/callgraph/VirtualCalls.java#L122?

Reply:
In https://github.com/Sable/soot/blob/fcc380a506258291d6fe6954425586a4ae4968ea/src/main/java/soot/jimple/toolkits/callgraph/VirtualCalls.java#L112, if a method is resolved by calling getMethodUnsafe(), and if the method is abstract we are not assigning it the method to ret. So ret would be null.
Whereas that's not the case in default branch.

Reply:
In default, we would look into the other interfaces for an implementation instead. Isn't that correct? After all, the interafcae implementatin hierarchy is kind of a tree where the method to resolve is the root. We traverse all paths from the root to find the closest (most specific) implementation of the method. If the method is abstract (not-implemented) in one interace, it might still be in another. Do I miss something here?

Reply:
Yeah, that's correct. But what if we did not find the implementation when traversed through the hierarchy. We are still returning the candidate which is an abstract method.

Reply:
Ahhh. Now I understand! Can you please fix this in your already opened merge request for the other fix?

Reply:
Wait, no. How can candidate ever be set to an abstract method?

Reply:
Its been assigned here at https://github.com/Sable/soot/blob/feature/default-methods/src/main/java/soot/FastHierarchy.java#L818

Reply:
Perfect! Please set `candidate` to `null` here: https://github.com/Sable/soot/blob/feature/default-methods/src/main/java/soot/FastHierarchy.java#L823

Reply:
Okay.